ui/components/src/EChart.tsx

```diff
@@ -121,62 +102,80 @@ export type OnEventsType<T> = {
 export interface EChartsProps<T> {
   option: EChartsCoreOption;
   sx?: SxProps<Theme>;
-  onChartInitialized?: (instance: ECharts) => void;
   onEvents?: OnEventsType<T>;
   _instance?: React.MutableRefObject<ECharts | undefined>;
-  parentContainer?: HTMLDivElement | null;
+  onChartInitialized?: (instance: ECharts) => void;
 }
 
-export function ECharts<T>({ sx, _instance, onChartInitialized, option, onEvents }: EChartsProps<T>) {
-  const [containerRef, setContainerRef] = useState<HTMLDivElement | null>(null);
-  const [echart, setChart] = useState<ECharts | undefined>(undefined);
-  // Create a chart instance in the container
-  useLayoutEffect(() => {
-    if (containerRef === null) return;
-
-    const chart = init(containerRef);
-    setChart(chart);
-    if (_instance !== undefined) {
-      _instance.current = chart;
+export const EChart = React.memo(function EChart<T>({
+  option,
+  sx,
+  onEvents,
+  _instance,
+  onChartInitialized,
+}: EChartsProps<T>) {
+  const prevOption = useRef<EChartsCoreOption | undefined>();
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const chartElement = useRef<ECharts | null>(null);
+
+  const initChartDom = useCallback(() => {
+    if (containerRef.current !== null && chartElement.current === null) {
+      chartElement.current = init(containerRef.current);
+      onChartInitialized?.(chartElement.current);
+      if (_instance !== undefined) {
+        _instance.current = chartElement.current;
+      }
     }
-    bindEvents(chart, onEvents);
+  }, [_instance, onChartInitialized]);
 
-    onChartInitialized?.(chart);
-
-    return () => {
-      chart.dispose();
-    };
-  }, [containerRef, _instance, onChartInitialized, onEvents]);
+  useEffect(() => {
+    const chart = chartElement.current;
+    if (chart === null || onEvents === undefined) return;
+    bindEvents(chart, onEvents);
+  }, [chartElement, onEvents]);
 
   // Sync options with chart instance
   useLayoutEffect(() => {
-    if (echart === undefined) return;
-    echart.setOption(option);
-  }, [echart, option]);
+    if (!isEqual(prevOption.current, option)) {
+      initChartDom();
+      const chart = chartElement.current;
+      if (chart === null) return;
+      chart.setOption(option, true);
+      prevOption.current = option;
+    }
+  }, [initChartDom, chartElement, option]);
 
+  // Resize chart, cleanup on unmount
   useLayoutEffect(() => {
     const updateSize = debounce(() => {
-      echart?.resize();
+      if (chartElement.current === null) return;
+      chartElement.current.resize();
     }, 200);
     window.addEventListener('resize', updateSize);
     updateSize();
-    return () => window.removeEventListener('resize', updateSize);
-  }, [echart]);
+    return () => {
+      window.removeEventListener('resize', updateSize);
+      if (chartElement.current !== null) {
+        chartElement.current.dispose();
+        chartElement.current = null;
+      }
+    };
```

Comment:
This feels weird to me because it's dispose function is cleaning up something from (I think?) another effect? Feels like each effect should be cleaning up after itself otherwise this is going to become really hard to reason about.

Reply:
Yeah good call, and now that event unbinding works, I was able to clean this up :tada: